Code/UnitTests/ToolsFoundationTest/FileSystem/FileSystemModelTest.cpp

```diff
@@ -0,0 +1,695 @@
+  ezMap<ezString, ezFileStatus::Status> referencedFolders;
+
+  EZ_TEST_BLOCK(ezTestBlock::Enabled, "Shutdown")
+  {
+    ezFileSystemModel::GetSingleton()->Deinitialize(&referencedFiles, &referencedFolders);
+    ezFileChangedEvent expected[] = {ezFileChangedEvent({}, {}, ezFileChangedEvent::Type::ModelReset)};
+    CompareFiles(ezMakeArrayPtr(expected));
+    ClearFiles();
+
+    ezFolderChangedEvent expected2[] = {ezFolderChangedEvent({}, ezFolderChangedEvent::Type::ModelReset)};
+    CompareFolders(ezMakeArrayPtr(expected2));
+    ClearFolders();
+  }
+
+  EZ_TEST_BLOCK(ezTestBlock::Enabled, "Startup Restore Model")
+  {
+    ezFileSystemModel::GetSingleton()->Initialize(fsConfig, std::move(referencedFiles), std::move(referencedFolders));
+
+    ezFileChangedEvent expected[] = {ezFileChangedEvent({}, {}, ezFileChangedEvent::Type::ModelReset)};
+    CompareFiles(ezMakeArrayPtr(expected));
+    ClearFiles();
+
+    ezFolderChangedEvent expected2[] = {ezFolderChangedEvent({}, ezFolderChangedEvent::Type::ModelReset)};
+    CompareFolders(ezMakeArrayPtr(expected2));
+    ClearFolders();
+
+    EZ_TEST_INT(ezFileSystemModel::GetSingleton()->GetFiles()->GetCount(), 1);
+    EZ_TEST_INT(ezFileSystemModel::GetSingleton()->GetFolders()->GetCount(), 2);
+  }
+
+  EZ_TEST_BLOCK(ezTestBlock::Enabled, "GetFiles")
+  {
+    ezStringBuilder sFilePathNew(sOutputFolder);
+    sFilePathNew.AppendPath("Folder2", "rootFile2.txt");
+
+    ezFileSystemModel::LockedFiles files = ezFileSystemModel::GetSingleton()->GetFiles();
+    EZ_TEST_INT(files->GetCount(), 1);
+    auto it = files->GetIterator();
+    EZ_TEST_STRING(it.Key(), sFilePathNew);
+    EZ_TEST_BOOL(it.Value().m_Timestamp.IsValid());
+    EZ_TEST_INT((ezInt64)it.Value().m_uiHash, (ezInt64)10983861097202158394u);
+    EZ_TEST_BOOL(it.Value().m_Status == ezFileStatus::Status::Valid);
+  }
+
+  EZ_TEST_BLOCK(ezTestBlock::Enabled, "GetFolders")
+  {
+    ezStringBuilder sFolder(sOutputFolder);
+    sFolder.AppendPath("Folder2");
+
+    ezFileSystemModel::LockedFolders folders = ezFileSystemModel::GetSingleton()->GetFolders();
+    EZ_TEST_INT(folders->GetCount(), 2);
+    auto it = folders->GetIterator();
+
+    // ezMap is sorted so the order is fixed.
+    EZ_TEST_STRING(it.Key(), sOutputFolder);
+    EZ_TEST_BOOL(it.Value() == ezFileStatus::Status::Valid);
+
+    it.Next();
+    EZ_TEST_STRING(it.Key(), sFolder);
+    EZ_TEST_BOOL(it.Value() == ezFileStatus::Status::Valid);
+  }
+
+  EZ_TEST_BLOCK(ezTestBlock::Enabled, "CheckFileSystem")
+  {
+    ezStringBuilder sFolderPath(sOutputFolder);
+    sFolderPath.AppendPath("Folder2");
+
+    ezStringBuilder sFilePath(sOutputFolder);
+    sFilePath.AppendPath("Folder2", "rootFile2.txt");
+
+    ezFileSystemModel::GetSingleton()->CheckFileSystem();
+
+    // #TODO_ASSET This FileChanged should be removed once the model is fixed to no longer require firing this after restoring the model from cache. See comment in ezFileSystemModel::HandleSingleFile.
```

Comment:
Yes. Right now this is a quirk that is required to make the asset curator work. Once it stores its entire state in the cache this will no longer be necessary.